### PR TITLE
TEAMROCKET#9: Improve Search Algorithm

### PR DESCRIPTION
**Pull request by williamcsimpson**
Only edited `search()` in `utils.js`. Implemented an improved search algorithm that ranks recipes by how closely they match the search query.

### CODE REVIEW

**Review by vchen51 — COMMENTED** [review]
nice : D

**Review by vchen51 — APPROVED** [review]

File: root/scripts/utils.js [hunk 1]
```diff
@@ -281,38 +281,64 @@ export function recipeIdArrayToObject(arr) {
...
+      let numMatchingTokens = 0;
+      let mostRecentMatch = tokenizedQuery.length;
+      let numMatchingCharacters = 0;
+      let searchScore = 0;
       const { title } = recipe;
       if (title) {
-        if (!title.toLowerCase().includes(query)) {
-          recipeMatches = false;
+        for (let j = tokenizedQuery.length - 1; j >= 0; j -= 1) {
+          if (title.toLowerCase().includes(tokenizedQuery[j])) {
```
Comment (lramos0) [inline root/scripts/utils.js:312]: This is a LOT more robust than what we had before. I was doing some testing and looks like "burger with garlic aioli" and "vegan pudding" work as search queries with this new algorithm. Where as, the old one did not display anything. I take it that matching each token (individual word) against the title is what allowed this.

For an even more robust algorithm, what if we also factored the `ingredients-list` as a factor in our search scores (like what we're doing with the title here). Let me know your thoughts, we can obviously open up a new ticket for this if time is short.

File: root/scripts/utils.js [hunk 2]
```diff
@@ -281,38 +281,64 @@ export function recipeIdArrayToObject(arr) {
...
-      if (searchResults.has(recipe)) {
-        searchResults.delete(recipe);
+      // remove recipe from results if an error thrown
+      if (searchResults.some((a) => a.recipe === recipe)) {
+        searchResults = searchResults.filter((a) => a.recipe !== recipe);
       }
     }
   }
-  return Array.from(searchResults);
+  searchResults.sort((a, b) => b.score - a.score); // put a before b if a.score > b.score
```
Comment (lramos0) [inline root/scripts/utils.js:339]: so clean, I like this a lot. The concept of search scores is obviously a lot more reliable than a simple binary true or false for title matching. Allows us to judge the quality of each result so users see best first, bonus 😄.

**Review by lramos0 — APPROVED** [review]
Amazing execution! You guys obviously gave it some thought, and created a smarter search algorithm than what we had before. Consider my comments about also using ingredients-list as a factor in generating search scores, other than that, should be good to go. 🙏

File: root/scripts/utils.js (hunk 1 above)
Comment (skmartinelli) [reply root/scripts/utils.js:312]: I think it'd be best to open a new ticket for ingredient search

**Pull request merged**